_posts/2020-12-01-configuring-and-suppressing-warnings.md

```diff
@@ -0,0 +1,192 @@
+Test.scala:4: warning: [deprecation @ my.app.C.f | origin=scala.Predef.ArrowAssoc.→ | version=2.13.0] method → in class ArrowAssoc is deprecated [...]
+  def f() = 1 → 2
+              ^
+```
+
+Here, the warning message includes the following additional information:
+
+* the warning category `deprecation`
+* the site where the warning is issued `my.app.C.f`
+* the origin of the deprecation `scala.Predef.ArrowAssoc.→`
+* the `since` version of the deprecation `2.13.0`
+
+### Filters
+
+The actions explained above are always applied to a set of warnings selected by a filter expression. The following filters are available:
+
+* `any` matches every message.
+* `cat=deprecation` filters according to the message category, for example deprecations (details below). 
+* `msg=regex` applies if some part of the message matches the regex.
+* `site=my\.package\.*` filters on the site where the warning is triggered. The regex must match the entity's full name (`package.Class.method`).
+* `src=src_managed/.*` filters warnings issued in a source file (details below).
+* Deprecation warnings can be filtered on two additional criteria:
+  * `origin=external\.package\..*` filters on full name of the deprecated entity.
+  * `since<1.24` filters on the `since` annotation argument of the deprecated entity (details below).
+
+Multiple filters can be combined using `&` to narrow down the selection. For example, the following configuration turns deprecation warnings for `scala.Predef` into errors:
+
+```scala
+$> scalac '-Wconf:cat=deprecation&origin=scala\.Predef\..*:e' Test.scala
+Test.scala:4: error: method → in class ArrowAssoc is deprecated [...]
+  def f() = 1 → 2
+              ^
+1 error
+```
+
+Note that the `-Wconf:...` compiler argument is between quotes (`'`) in the command line, which prevents the shell from interpreting characters like `&` or `*`.
+
+For some of the filters the syntax is not trivial, so we look at them in more detail.
+
+* **Message category**: Every message has a category that is displayed in verbose mode (`-Wconf:any:wv`). The `-Wconf:help` option displays the full list of available categories. For example, every `-Xlint` warning has its own category (`lint-infer-any`), the super-category `lint` matches all lint warnings.
+* **Source file**: By default, the source file filter is a regex that must match the file path relative to any path segment. For example, `b/.*Test.scala` matches `/a/b/XTest.scala` but not `/ab/Test.scala`. If the `-rootdir` compiler option is specified, the regex must match the file path relative to that root directory.
+* **Since version for deprecations**: In a `since<1.24` filter expression, valid operators are `<` `=` and `>` and valid version numbers are `N`, `N.M` and `N.M.P`. Because the `since` anontation argument can contain arbitrary text, the first version number found in the text is used for the comparison, for example `1.2.3` in `@deprecated("", "some lib 1.2.3-foo")`.
```

Comment:
```suggestion
* **Since version for deprecations**: In a `since<1.24` filter expression, valid operators are `<`, `=` and `>` and valid version numbers are `N`, `N.M` and `N.M.P`. Because the `since` annotation argument can contain arbitrary text, the first version number found in the text is used for the comparison, for example `1.2.3` in `@deprecated("", "some lib 1.2.3-foo")`.
```